core/metacling/src/TCling.cxx

```diff
@@ -4191,19 +4194,27 @@ TCling::CheckClassInfo(const char *name, Bool_t autoload, Bool_t isClassOrNamesp
    // this forward declaration.
    const cling::LookupHelper& lh = fInterpreter->getLookupHelper();
    const clang::Type *type = nullptr;
-   const clang::Decl *decl
+   decl
       = lh.findScope(classname,
                      gDebug > 5 ? cling::LookupHelper::WithDiagnostics
                      : cling::LookupHelper::NoDiagnostics,
-                     &type, /* intantiateTemplate= */ false );
+                     &type, instantiateTemplate);
```

Comment:
To instantiate or not to instantiate is a very delicate choice.  Any change needs to be carefully described in at least the git log.

What does this change addresses?  What was not working or what new behavior/feature requires this?  Why was the previous choice (you may want to review the git history of the line) incorrect?

Reply:
mmm the reason is that there are divergences between the code called by TCling::CheckClassInfo() and the code called by TClingClassInfo constructor. If we want to reuse the first lookup for speeding up the second, I thought they need to be equalized otherwise we cannot reuse the decl to obtain the same result.

See diff:
![image](https://github.com/root-project/root/assets/10653970/5688df2f-8408-4776-82e3-9a4f479405de)

Reply:
> , I thought they need to be equalized 

They can not be equalized without understanding precise why they are different.  If I recall correctly they do thing differently on purpose, `CheckClassInfo` checks gently on what exists while `TClingClassInfo` must create it if can be created but is not yet.

So if the decl was found by `CheckClassInfo` then it can be passed to `TClingClassInfo` to use otherwise .... well their should not be an otherwise, if `CheckClassInfo` did not find it, it should not be creating a `TClingClassInfo` ... (but `TClingClassInfo` can and will be called from other places without a decl).

Reply:
I'm not sure how to proceed now. Maybe it's just not possible to replace the two lookups by one of they are slightly different?

Reply:
Let's first try to remove the new code that is changing the instantiation behavior.
```suggestion
                     &type);
```